File: _designs_001/Read N Characters Given Read4 II/ReaderN.java
public class ReaderN {
    private Reader4 reader4;
    private int offset, bufsize;

    ReaderN(Reader4 reader4) {
        this.reader4 = reader4;
        this.offset = 0;
        this.bufsize = 0;
    }

    private char[] buffer = new char[4];

    public int read(char[] buf, int n) {
        int readBytes = 0;
        boolean eof = false;
        while (!eof && readBytes < n) {
            int sz = (bufsize > 0) ? bufsize : reader4.read(buffer);
            if (bufsize == 0 && sz < 4)
                eof = true;
            int bytes = Math.min(n - readBytes, sz);
            System.arraycopy(buffer, offset, buf, readBytes, bytes);
            offset = (offset + bytes) % 4;
            bufsize = sz - bytes;
            readBytes += bytes;
        }
        return readBytes;
    }
}

File: _designs_001/Read N Characters Given Read4 II/ReaderN.rb
class ReaderN
  def initialize(reader4)
    @reader4 = reader4
    @left_over = ''
  end

  def read(buf, n)
    buf.delete!(buf)
    box4, eof = '', false
    while (!eof and @left_over.length < n)
      box_size = @reader4.read(box4)
      eof = true if box_size < 4
      @left_over << box4
    end

    size = [@left_over.length, n].min
    buf << @left_over[0...size]
    @left_over.slice!(0...size)
    buf.length
  end
end
